RetailCoder.VBE/Navigation/CodeMetrics/CodeMetricsAnalyst.cs

```diff
@@ -0,0 +1,194 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Rubberduck.Parsing.VBA;
+using Antlr4.Runtime.Tree;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.VBEditor;
+using Antlr4.Runtime.Misc;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.SmartIndenter;
+
+namespace Rubberduck.Navigation.CodeMetrics
+{
+    public class CodeMetricsAnalyst : ICodeMetricsAnalyst
+    {
+        private readonly IIndenterSettings indenterSettings;
+    
+        public CodeMetricsAnalyst(IIndenterSettings indenterSettings)
+        {
+            this.indenterSettings = indenterSettings;
+        }
+
+        public IEnumerable<ModuleMetricsResult> ModuleMetrics(RubberduckParserState state)
+        {
+            if (state == null || !state.AllUserDeclarations.Any())
+            {
+                // must not return Enumerable.Empty
+                yield break;
+            }
+
+            var trees = state.ParseTrees;
+            var results = new List<CodeMetricsResult>();
+
+            foreach (var moduleTree in trees)
+            {
+                yield return GetModuleResult(moduleTree.Key, moduleTree.Value, state.DeclarationFinder);
+            };
+        }
+
+        public ModuleMetricsResult GetModuleResult(RubberduckParserState state, QualifiedModuleName qmn)
+        {
+            return GetModuleResult(qmn, state.GetParseTree(qmn), state.DeclarationFinder);
+        }
+
+        private ModuleMetricsResult GetModuleResult(QualifiedModuleName qmn, IParseTree moduleTree, DeclarationFinder declarationFinder)
+        {
+            // Consider rewrite as visitor? That should make subtrees easier and allow us to expand metrics
+            var cmListener = new CodeMetricsListener(declarationFinder, indenterSettings);
+            ParseTreeWalker.Default.Walk(cmListener, moduleTree);
+            return cmListener.GetMetricsResult(qmn);
+        }
+
+
+        private class CodeMetricsListener : VBAParserBaseListener
```

Comment:
Why is there only one listener for all three types of code metrics? Wouldn't it be better to have one for each type and then just combine them for the walk over the tree?

Reply:
Same reasoning as above. In principle the idea is correct, it seemed like too much of a hassle for something we'd probably not need.